File: ui/apps/everest/.e2e/utils/monitoring-instance.ts
// everest
// Copyright (C) 2023 Percona LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { APIRequestContext, expect } from '@playwright/test';
import { execSync } from 'child_process';
import { getDBClientPod } from '@e2e/utils/db-cmd-line';
import { getK8sUid } from '@e2e/utils/kubernetes';

const {
  MONITORING_URL,
  MONITORING_USER,
  MONITORING_PASSWORD
} = process.env;

export const testMonitoringName = 'ui-test-monitoring';
export const testMonitoringName2 = 'ui-test-monitoring-1';

export const createMonitoringInstance = async (
  request: APIRequestContext,
  name: string,
  namespace: string,
  token: string
) => {
  const data = {
    name,
    type: 'pmm',
    url: MONITORING_URL,
    allowedNamespaces: [],
    verifyTLS: false,
    pmm: {
      user: MONITORING_USER,
      password: MONITORING_PASSWORD,
    },
  };

  const response = await request.post(
    `/v1/namespaces/${namespace}/monitoring-instances`,
    {
      data,
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );

  expect(response.ok()).toBeTruthy();
};

export const deleteMonitoringInstance = async (
  request: APIRequestContext,
  namespace,
  name,
  token: string
) => {
  const response = await request.delete(
    `/v1/namespaces/${namespace}/monitoring-instances/${name}`,
    {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );
  expect(response.ok()).toBeTruthy();
};

export const listMonitoringInstances = async (
  request: APIRequestContext,
  namespace,
  token: string
) => {
  const response = await request.get(
    `/v1/namespaces/${namespace}/monitoring-instances`,
    {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );
  expect(response.ok()).toBeTruthy();

  const responseBody = await response.json();
  return responseBody;
};

export const checkDBMetrics = async (
  metric: string,
  instance: string,
  userPass: string
): Promise<void> => {
  const clientPod = await getDBClientPod('psmdb', 'db-client');
  const date = new Date();
  const end = Math.floor(date.getTime() / 1000);
  const start = end - 60;
  const endpoint = `monitoring-service.everest-system:443`;
  const url = `https://${userPass}@${endpoint}/graph/api/datasources/proxy/1/api/v1/query_range?query=min%28${metric}%7Bnode_name%3D%7E%22${instance}%22%7d%20or%20${metric}%7Bnode_name%3D%7E%22${instance}%22%7D%29&start=${start}&end=${end}&step=60`;

  try {
    const command = `kubectl exec --namespace db-client ${clientPod} -- curl -s -k "${url}"`;
    var output = JSON.parse(execSync(command).toString());
  } catch (error) {
    console.error(`Error executing command: ${error}`);
    throw error;
  }

  const metrics = parseInt(output?.data?.result[0]?.values[0][1]);
  expect(metrics).toBeDefined();
  expect(typeof metrics).toBe('number');
  expect(Number.isNaN(metrics)).toBeFalsy();
};

export const checkK8sMetrics = async (
  metric: string,
  userPass: string
): Promise<void> => {
  const clientPod = await getDBClientPod('psmdb', 'db-client');
  const k8sUid = await getK8sUid();
  const date = new Date();
  const end = Math.floor(date.getTime() / 1000);
  const start = end - 60;
  const endpoint = 'monitoring-service.everest-system:443';
  const url = `https://${userPass}@${endpoint}/graph/api/datasources/proxy/1/api/v1/query_range?query=count%28${metric}%7Bk8s_cluster_id%3D%22${k8sUid}%22%7D%29&start=${start}&end=${end}&step=60`;

  try {
    const command = `kubectl exec --namespace db-client ${clientPod} -- curl -s -k "${url}"`;
    var output = JSON.parse(execSync(command).toString());
  } catch (error) {
    console.error(`Error executing command: ${error}`);
    throw error;
  }

  const metrics = parseInt(output?.data?.result[0]?.values[0][1]);
  expect(metrics).toBeDefined();
  expect(typeof metrics).toBe('number');
  expect(Number.isNaN(metrics)).toBeFalsy();
};

export const checkQAN = async (
  serviceType: string,
  node: string,
  userPass: string
): Promise<void> => {
  const clientPod = await getDBClientPod('psmdb', 'db-client');
  const now = new Date();
  const start = new Date(now.getTime() - 12 * 60 * 60 * 1000).toISOString();
  const end = now.toISOString();

  const endpoint = 'monitoring-service.everest-system:443';
  const url = `https://${userPass}@${endpoint}/v0/qan/ObjectDetails/GetMetrics`;

  const payload = JSON.stringify({
    period_start_from: start,
    period_start_to: end,
    group_by: 'queryid',
    filter_by: node,
    labels: [
      { key: 'service_type', value: [serviceType] },
      { key: 'node_name', value: [node] },
    ],
    limit: 10,
    totals: true,
  });

  const encodedPayload = Buffer.from(payload).toString('base64');

  try {
    const command = `kubectl exec -n db-client ${clientPod} -- sh -c "echo '${encodedPayload}' | base64 -d > /tmp/payload.json && curl -s -k -XPOST -d @/tmp/payload.json '${url}'"`;
    var output = JSON.parse(execSync(command).toString());
  } catch (error) {
    console.error(`Error executing command: ${error}`);
    throw error;
  }

  // we use node_name because if it's returned it means there is some data for the node
  const nodeName = output?.metadata?.node_name.toString();
  expect(nodeName).toBeTruthy();
  expect(nodeName?.trim()).not.toBe('');
};

File: internal/server/monitoring_instance.go
// everest
// Copyright (C) 2023 Percona LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package server

import (
	"net/http"

	"github.com/labstack/echo/v4"

	"github.com/percona/everest/api"
)

// CreateMonitoringInstance creates a new monitoring instance.
func (e *EverestServer) CreateMonitoringInstance(ctx echo.Context, cluster, namespace string) error {
	var params api.CreateMonitoringInstanceJSONRequestBody
	if err := ctx.Bind(&params); err != nil {
		return err
	}

	created, err := e.handler.CreateMonitoringInstance(ctx.Request().Context(), cluster, namespace, &params)
	if err != nil {
		return err
	}

	out := &api.MonitoringInstance{}
	out.FromCR(created)
	return ctx.JSON(http.StatusOK, out)
}

// ListMonitoringInstances lists all monitoring instances.
func (e *EverestServer) ListMonitoringInstances(ctx echo.Context, cluster, namespace string) error {
	mcList, err := e.handler.ListMonitoringInstances(ctx.Request().Context(), cluster, namespace)
	if err != nil {
		return err
	}

	result := make([]*api.MonitoringInstance, 0, len(mcList.Items))
	for _, mc := range mcList.Items {
		out := &api.MonitoringInstance{}
		out.FromCR(&mc)
		result = append(result, out)
	}
	return ctx.JSON(http.StatusOK, result)
}

// GetMonitoringInstance retrieves a monitoring instance.
func (e *EverestServer) GetMonitoringInstance(ctx echo.Context, cluster, namespace, name string) error {
	m, err := e.handler.GetMonitoringInstance(ctx.Request().Context(), cluster, namespace, name)
	if err != nil {
		return err
	}

	out := &api.MonitoringInstance{}
	out.FromCR(m)
	return ctx.JSON(http.StatusOK, out)
}

// UpdateMonitoringInstance updates a monitoring instance based on the provided fields.
func (e *EverestServer) UpdateMonitoringInstance(ctx echo.Context, cluster, namespace, name string) error {
	var params api.UpdateMonitoringInstanceJSONRequestBody
	if err := ctx.Bind(&params); err != nil {
		return err
	}

	updated, err := e.handler.UpdateMonitoringInstance(ctx.Request().Context(), cluster, namespace, name, &params)
	if err != nil {
		return err
	}

	out := &api.MonitoringInstance{}
	out.FromCR(updated)
	return ctx.JSON(http.StatusOK, out)
}

// DeleteMonitoringInstance deletes a monitoring instance.
func (e *EverestServer) DeleteMonitoringInstance(ctx echo.Context, cluster, namespace, name string) error {
	if err := e.handler.DeleteMonitoringInstance(ctx.Request().Context(), cluster, namespace, name); err != nil {
		return err
	}

	return ctx.NoContent(http.StatusNoContent)
}
